src/webgpu/shader/validation/shader_io/builtins.spec.ts

```diff
@@ -0,0 +1,131 @@
+export const description = `Validation tests for entry point built-in variables`;
+
+import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { ShaderValidationTest } from '../shader_validation_test.js';
+
+import { generateShader } from './util.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+// List of all built-in variables and their stage, in|out usage, and type.
+// Taken from table in Section 15:
+// https://www.w3.org/TR/2021/WD-WGSL-20211013/#builtin-variables
+const kBuiltins = [
+  { name: 'vertex_index', stage: 'vertex', io: 'in', type: 'u32' },
+  { name: 'instance_index', stage: 'vertex', io: 'in', type: 'u32' },
+  { name: 'position', stage: 'vertex', io: 'out', type: 'vec4<f32>' },
+  { name: 'position', stage: 'fragment', io: 'in', type: 'vec4<f32>' },
+  { name: 'front_facing', stage: 'fragment', io: 'in', type: 'bool' },
+  { name: 'local_invocation_id', stage: 'compute', io: 'in', type: 'vec3<u32>' },
+  { name: 'local_invocation_index', stage: 'compute', io: 'in', type: 'u32' },
+  { name: 'global_invocation_id', stage: 'compute', io: 'in', type: 'vec3<u32>' },
+  { name: 'workgroup_id', stage: 'compute', io: 'in', type: 'vec3<u32>' },
+  { name: 'num_workgroups', stage: 'compute', io: 'in', type: 'vec3<u32>' },
+  { name: 'sample_index', stage: 'fragment', io: 'in', type: 'u32' },
+  { name: 'sample_mask', stage: 'fragment', io: 'in', type: 'u32' },
+  { name: 'sample_mask', stage: 'fragment', io: 'out', type: 'u32' },
+] as const;
+
+// List of types to test against.
+const kTestTypes = [
+  'bool',
+  'u32',
+  'i32',
+  'f32',
+  'vec2<bool>',
+  'vec2<u32>',
+  'vec2<i32>',
+  'vec2<f32>',
+  'vec3<bool>',
+  'vec3<u32>',
+  'vec3<i32>',
+  'vec3<f32>',
+  'vec4<bool>',
+  'vec4<u32>',
+  'vec4<i32>',
+  'vec4<f32>',
+  'mat2x2<f32>',
+  'mat2x3<f32>',
+  'mat2x4<f32>',
+  'mat3x2<f32>',
+  'mat3x3<f32>',
+  'mat3x4<f32>',
+  'mat4x2<f32>',
+  'mat4x3<f32>',
+  'mat4x4<f32>',
+  'atomic<u32>',
+  'atomic<i32>',
+  'array<bool,4>',
+  'array<u32,4>',
+  'array<i32,4>',
+  'array<f32,4>',
+  'MyStruct',
+] as const;
+
+g.test('stage_inout')
+  .desc(`Test validation of built-in variable stage and in/out usage`)
```

Comment:
Go into slightly more detail here so it's possible to understand the test without reading it, e.g. "Test different [[builtin]] declarations are validated to match the stage and in/out usage of the specified builtin."